### ISSUE TITLE
Confused about contrastive_loss

### ISSUE THREAD

**Issue by BrandonHanx** [TencentYoutuResearch/PersonReID-NAFS#5]
@girafffeee Thanks for your work!
However, I am confused about what this loss function to do:
https://github.com/TencentYoutuResearch/PersonReID-NAFS/blob/71bf192f464a410845a88590e4660eeeba1a1e50/utils/metric.py#L265-L286

Why the `criterion `and `t2i_similarities `are not used in this function?
Does this function do the same thing like CSAL in the paper?

**Comment by wettera** [comment]
> @girafffeee Thanks for your work!
> However, I am confused about what this loss function to do:
> https://github.com/TencentYoutuResearch/PersonReID-NAFS/blob/71bf192f464a410845a88590e4660eeeba1a1e50/utils/metric.py#L265-L286
> 
> Why the `criterion `and `t2i_similarities `are not used in this function?
> Does this function do the same thing like CSAL in the paper?

Thanks for your attention on our work. `criterion` is not used, just ignore it. We compute loss in `constrastive_loss = torch.mean(torch.sum(i2t_loss, dim=1))`. As for the `t2i_similarities`, experimental results show that only use `i2t_similarities` can achieve higher performance.

**Comment by BrandonHanx** [comment]
Hi, @wettera . Many thanks for your reply!

Would you please tell me what is the corresponding formula for line 276-278.
It seems like used another temperature softmax on S(I, T) in the formula 6 in the paper.

Please point out if I am wrong.

**Comment by wettera** [comment]
Yes, line 276-278 is related to L_i in formula 6. S(I, T) = ` i2t_pred = F.softmax(i2t_similarites * self.args.lambda_softmax, dim=1)`.
log(S(I,T)/(q+ $\epsilon$)) = `(F.log_softmax(i2t_similarites * self.args.lambda_softmax, dim=1) - torch.log(labels_mask_norm + self.epsilon))`

**Comment by BrandonHanx** [comment]
Yes, I agree with your aforementioned explanation.
However, my question is why S(I, T) = `i2t_pred = F.softmax(i2t_similarites * self.args.lambda_softmax, dim=1)`?
According to formula 5, maybe S(I, T) = `i2t_similarites `?

**Comment by wettera** [comment]
I understand. I think adding softmax function is beneficial for training. The loss function may converge faster. We will update arxiv later. Thanks very much.

**Comment by BrandonHanx** [comment]
Ok. Thanks for your patience.
Because there are some differences from this code and your paper, would you please report more detailed and solid results in you updated arxiv? I am really appreciate it.

BTW, the \lamda2 for softmax and another \lamda2 for loss balance would make some ambiguity. Please take a look.
